src/bootstrap/config.rs

```diff
@@ -1910,11 +1939,28 @@ impl Config {
 
         // Look for a version to compare to based on the current commit.
         // Only commits merged by bors will have CI artifacts.
+
+        // If we are running in CI, the current commit will never have artifacts already built.
+        let tip_commit = if CiEnv::is_ci() { "HEAD^" } else { "HEAD" };
+        // NOTE: we intentionally don't filter compiler/ and library/ here, even though we ideally
+        // would to avoid unnecessary downloads. The reason for that is complicated:
+        // - When bors creates a merge commit for a rollup, there are two levels of indirection:
+        //   1. The merge by bors ("Auto merge of...")
+        //   2. The merge from each individual commit in the rollup ("Rollup merge of ...")
+        // - Because of the indirection, `git log --author=bors compiler/` *only* includes merges that weren't from a rollup.
+        // - To avoid filtering out rollup merges, we need to do something like
+        //   1. Find the last commit that modified compiler/: CHANGE
+        //   2. Find the last commit authored by bors (if in CI, excluding the current commit): BORS
+        //   3. Confirm that CHANGE is in the history of BORS
+        //
+        // That's hard both to implement and to test. For now, we just unconditionally use the last
```

Comment:
Do we not accomplish this with `git log --author=bors --first-parent -n1 tip_commit -- compiler`?

In any case, I guess this is commentary that does not strictly apply to this PR - maybe we can move iterating on it to an issue or something and drop it from this PR. (As you suggested)

Reply:
i opened https://github.com/rust-lang/rust/issues/113250 and removed the comment.